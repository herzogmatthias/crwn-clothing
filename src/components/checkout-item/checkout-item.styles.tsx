import styled, { css } from "styled-components";

const checkoutSpanwidth = css`
  width: 23%;
`;

export const CheckoutItemContainer = styled.div`
  width: 100%;
  display: flex;
  min-height: 100px;
  border-bottom: 1px solid darkgrey;
  padding: 15px 0;
  font-size: 20px;
  align-items: center;
`;

export const ImageContainer = styled.div`
  width: 23%;
  padding-right: 15px;
`;
export const CheckoutItemImage = styled.img`
  width: 100%;
  height: 100%;
`;
const [NameContainer, PriceContainer] = Array(2).fill(
  styled.span`
    ${checkoutSpanwidth}
  `
);
export const QuantityContainer = styled.span`
  ${checkoutSpanwidth}
  display: flex;
`;

export { PriceContainer, NameContainer };

export const ArrowContainer = styled.div`
  cursor: pointer;
`;
export const ValueContainer = styled.span`
  margin: 0 10px;
`;
export const RemoveButtonContainer = styled.div`
  padding-left: 12px;
  cursor: pointer;
`;
