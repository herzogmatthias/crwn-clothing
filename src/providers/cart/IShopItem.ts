export interface IShopItem {
  id: number;
  name: string;
  imageUrl: string;
  price: number;
}
