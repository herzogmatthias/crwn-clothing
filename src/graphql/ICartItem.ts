import { IShopItem } from "./IShopItem";

export interface ICartItem extends IShopItem {
  quantity: number;
}
