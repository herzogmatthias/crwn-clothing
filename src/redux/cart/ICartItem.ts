import { IShopItem } from "../shop/IShopItem";

export interface ICartItem extends IShopItem {
  quantity: number;
}
