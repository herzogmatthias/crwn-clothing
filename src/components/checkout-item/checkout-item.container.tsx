import { graphql, MutationFunctionOptions } from "react-apollo";
import React from "react";
import { flowRight } from "lodash";
import {
  TOGGLE_CART_HIDDEN,
  CLEAR_ITEM_FROM_CART,
  ADD_ITEM_TO_CART,
  REMOVE_ITEM_FROM_CART,
} from "../../graphql/mutations";
import { ApolloCurrentQueryResult } from "apollo-boost";
import { GET_ITEM_COUNT } from "../../graphql/queries";
import { ICartItem } from "../../graphql/ICartItem";
import { IShopItem } from "../../graphql/IShopItem";
import CheckoutItem from "./checkout-item.component";

interface IWithGraphqlProps {
  clearItemFromCart(
    options: MutationFunctionOptions<void, { item: ICartItem }>
  ): void;
  addItem(options: MutationFunctionOptions<void, { item: ICartItem }>): void;
  removeItem(options: MutationFunctionOptions<void, { item: ICartItem }>): void;
  cartItem: ICartItem;
}

const withGraphql = ({
  clearItemFromCart,
  cartItem,
  addItem,
  removeItem,
}: IWithGraphqlProps) => {
  return (
    <CheckoutItem
      cartItem={cartItem}
      clearItemFromCart={() =>
        clearItemFromCart({ variables: { item: cartItem } })
      }
      addItem={() => addItem({ variables: { item: cartItem } })}
      removeItem={() => removeItem({ variables: { item: cartItem } })}
    ></CheckoutItem>
  );
};

export default flowRight(
  graphql<{ item: ICartItem }, { clearItemFromCart(item: ICartItem): void }>(
    CLEAR_ITEM_FROM_CART,
    { name: "clearItemFromCart" }
  ),
  graphql<{ item: IShopItem }, { addItemToCart(item: IShopItem): void }>(
    ADD_ITEM_TO_CART,
    {
      name: "addItem",
    }
  ),
  graphql<{ item: ICartItem }, { removeItemFromCart(item: ICartItem): void }>(
    REMOVE_ITEM_FROM_CART,
    {
      name: "removeItem",
    }
  )
)(withGraphql);
