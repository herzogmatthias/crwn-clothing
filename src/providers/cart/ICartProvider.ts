import { ICartItem } from "../../redux/cart/ICartItem";
import { IShopItem } from "../../redux/shop/IShopItem";

export interface ICartProvider {
  hidden: boolean;
  total: number;
  toggleHidden(): void;
  cartItems: ICartItem[];
  addItem(item: IShopItem): void;
  removeItem(item: ICartItem): void;
  clearItemFromCart(item: ICartItem): void;
  cartItemsCount: number;
}
