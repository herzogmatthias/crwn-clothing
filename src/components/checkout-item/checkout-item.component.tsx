import * as React from "react";

import "./checkout-item.styles.scss";
import { ICartItem } from "../../redux/cart/ICartItem";
import { ConnectedProps, connect } from "react-redux";
import {
  clearItemFromCart,
  addItem,
  removeItem,
} from "../../redux/cart/cart.actions";

type ICheckoutItemProps = ConnectedProps<typeof connector> & {
  cartItem: ICartItem;
};

function CheckoutItem({
  cartItem,
  clearItemFromCart,
  addItem,
  removeItem,
}: ICheckoutItemProps) {
  const { name, price, imageUrl, quantity } = cartItem;
  return (
    <div className="checkout-item">
      <div className="image-container">
        <img src={imageUrl} alt="item"></img>
      </div>
      <span className="name">{name}</span>
      <span className="quantity">
        <div className="arrow" onClick={() => removeItem(cartItem)}>
          &#10094;
        </div>
        <span className="value">{quantity}</span>
        <div className="arrow" onClick={() => addItem(cartItem)}>
          &#10095;
        </div>
      </span>
      <span className="price">{price}</span>
      <div
        onClick={() => clearItemFromCart(cartItem)}
        className="remove-button"
      >
        &#10005;
      </div>
    </div>
  );
}

const mapDispatchToProps = {
  clearItemFromCart: clearItemFromCart,
  addItem: addItem,
  removeItem: removeItem,
};
const connector = connect(null, mapDispatchToProps);
export default connect(null, mapDispatchToProps)(CheckoutItem);
