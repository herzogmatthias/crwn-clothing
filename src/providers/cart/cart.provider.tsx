import React, { createContext, useEffect } from "react";
import { ICartProvider } from "./ICartProvider";
import { ICartItem } from "../../redux/cart/ICartItem";
import {
  addItemToCart,
  removeItemFromCart,
  filterItemFromCart,
  getCartItemsCount,
  getTotalPrice,
} from "../../redux/cart/cart.utils";
import { IShopItem } from "../../redux/shop/IShopItem";

export const CartContext = createContext<ICartProvider>({
  hidden: true,
  total: 0,
  cartItems: [],
  cartItemsCount: 0,
  toggleHidden: () => {},
  addItem: () => {},
  removeItem: () => {},
  clearItemFromCart: () => {},
});

interface ICartProviderProps {
  children: React.ReactNode;
}

const CartProvider = ({ children }: ICartProviderProps) => {
  const [hidden, setHidden] = React.useState(true);

  const [cartItems, setCartItems] = React.useState<ICartItem[]>([]);
  const [cartItemsCount, setCartItemsCount] = React.useState(0);

  const [total, setTotal] = React.useState(0);

  const toggleHidden = () => {
    setHidden(!hidden);
  };

  const addItem = (item: IShopItem) =>
    setCartItems(addItemToCart(cartItems, item));

  const removeItem = (item: ICartItem) =>
    setCartItems(removeItemFromCart(cartItems, item));
  const clearItemFromCart = (item: ICartItem) =>
    setCartItems(filterItemFromCart(cartItems, item));

  useEffect(() => {
    setCartItemsCount(getCartItemsCount(cartItems));
    setTotal(getTotalPrice(cartItems));
  }, [cartItems]);

  return (
    <CartContext.Provider
      value={{
        total,
        addItem,
        hidden,
        toggleHidden,
        cartItems,
        cartItemsCount,
        removeItem,
        clearItemFromCart,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};

export default CartProvider;
