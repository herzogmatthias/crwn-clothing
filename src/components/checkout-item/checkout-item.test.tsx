import React from "react";
import { CheckoutItem } from "./checkout-item.component";
import { ActionCreatorWithPayload } from "@reduxjs/toolkit";
import { ICartItem } from "../../redux/cart/ICartItem";
import {
  REMOVE_ITEM,
  ADD_ITEM,
  CLEAR_ITEM_FROM_CART,
} from "../../redux/cart/cart.types";
import { ShallowWrapper, shallow } from "enzyme";
import { ArrowContainer, RemoveButtonContainer } from "./checkout-item.styles";

describe("NON CONNECTED CHECKOUT-ITEM TESTS", () => {
  let wrapper: ShallowWrapper;
  let mockAddItem: ActionCreatorWithPayload<ICartItem, "ADD_ITEM">;
  let mockRemoveItem: ActionCreatorWithPayload<ICartItem, "REMOVE_ITEM">;
  const dispatch = jest.fn();
  let mockClearItemFromCart: ActionCreatorWithPayload<
    ICartItem,
    "CLEAR_ITEM_FROM_CART"
  >;
  const cartItem: ICartItem = {
    id: 1,
    imageUrl: "asdfasdf",
    name: "hats",
    price: 23,
    quantity: 2,
  };

  beforeEach(() => {
    mockAddItem = jest.fn((item: ICartItem) => {
      dispatch({ type: ADD_ITEM, payload: item });
    }) as any;
    mockRemoveItem = jest.fn((item: ICartItem) => {
      dispatch({ type: REMOVE_ITEM, payload: item });
    }) as any;
    mockClearItemFromCart = jest.fn((item: ICartItem) => {
      dispatch({ type: CLEAR_ITEM_FROM_CART, payload: item });
    }) as any;

    const mockProps = {
      addItem: mockAddItem,
      clearItemFromCart: mockClearItemFromCart,
      removeItem: mockRemoveItem,

      cartItem: cartItem,
    } as any;
    wrapper = shallow(<CheckoutItem {...mockProps}></CheckoutItem>);
  });
  test("render Checkout-Item Component", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("dispatch AddItem when right arrow is clicked", () => {
    wrapper.find(ArrowContainer).at(1).simulate("click");
    expect(mockAddItem).toHaveBeenCalledWith(cartItem);
  });
  test("dispatch removeItem when left arrow is clicked", () => {
    wrapper.find(ArrowContainer).at(0).simulate("click");
    expect(mockRemoveItem).toHaveBeenCalledWith(cartItem);
  });
  test("dispatch clearItemFromCart when Remove Button is clicked", () => {
    wrapper.find(RemoveButtonContainer).simulate("click");
    expect(mockClearItemFromCart).toHaveBeenCalledWith(cartItem);
  });
});
