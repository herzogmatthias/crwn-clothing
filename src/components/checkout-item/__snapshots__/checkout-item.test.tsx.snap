// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NON CONNECTED CHECKOUT-ITEM TESTS render Checkout-Item Component 1`] = `
<styled.div>
  <styled.div>
    <styled.img
      alt="item"
      src="asdfasdf"
    />
  </styled.div>
  <styled.span>
    hats
  </styled.span>
  <styled.span>
    <styled.div
      onClick={[Function]}
    >
      ❮
    </styled.div>
    <styled.span>
      2
    </styled.span>
    <styled.div
      onClick={[Function]}
    >
      ❯
    </styled.div>
  </styled.span>
  <styled.span>
    23
  </styled.span>
  <styled.div
    onClick={[Function]}
  >
    ✕
  </styled.div>
</styled.div>
`;
